### PR TITLE
Allow further configuration of Elasticsearch RestClient through HttpClientConfigCallback

### LINKED ISSUE #14754: Allow further configuration of Elasticsearch RestClient through custom HttpClientConfigCallback
**Description**
Please consider providing more control over the configuration of the Elasticsearch RestClient with a user pluggable HttpClientConfigCallback implementation in the elasticsearch-rest-client extension. 

**Why**
Form my case, I need to set a SSLContext (as described [here](https://www.elastic.co/guide/en/elasticsearch/client/java-rest/current/_encrypted_communication.html)) to enable me to use self signed certificates generated by Elasticsearch kubernetes operator with a native image build. Currently one needs to provide these certificates at native image build time which is kind of more difficult and hard to maintain and much less flexible. 

**Implementation ideas**
One way to achieve this is to have a no op (i.e. empty), @Singleton scope @DefaultBean implementing HttpClientConfigCallback and inject a HttpClientConfigCallback in the ElasticsearchRestClientProducer and pass it along with the ElasticsearchConfig parameter to the RestClientBuilderHelper. createRestClientBuilder method. ElasticsearchRestClientProducer.createRestClientBuilder method than can call this instance on top of its current configuration. Since DefaultBean implementation does nothing, this will change nothing for anyone. But if a user provides an implementation, overriding the noop default, they can have the opportunity to provide custom configuration.

### PR DESCRIPTION
**Pull request by eAybars**
fixes #14754

### CODE REVIEW

File: extensions/elasticsearch-rest-client/runtime/src/main/java/io/quarkus/elasticsearch/restclient/lowlevel/runtime/RestClientBuilderHelper.java
Comment (loicmathieu) [inline extensions/elasticsearch-rest-client/runtime/src/main/java/io/quarkus/elasticsearch/restclient/lowlevel/runtime/RestClientBuilderHelper.java:85]: This line didn't improve readibility, can you get rid of it ?

File: extensions/elasticsearch-rest-client/runtime/src/main/java/io/quarkus/elasticsearch/restclient/lowlevel/runtime/RestClientBuilderHelper.java
Comment (loicmathieu) [inline extensions/elasticsearch-rest-client/runtime/src/main/java/io/quarkus/elasticsearch/restclient/lowlevel/runtime/RestClientBuilderHelper.java:88]: Can you split this line in two to improve readibility ? I have a hard time understanding what it is about at first read.
```suggestion
                List<InstanceHandle<RestClientBuilder.HttpClientConfigCallback>> handles =  Arc.container()
                        .select(RestClientBuilder.HttpClientConfigCallback.class, new ElasticsearchClientConfig.Literal())
                        .handles();
                for (InstanceHandle<RestClientBuilder.HttpClientConfigCallback> handle : handles) {
```

**Review by loicmathieu — COMMENTED** [review]
If I understand it correctly, this allow to add `HttpClientConfigCallback` that will be used inside our own `HttpClientConfigCallback` to further configure the HttpClient.

It's missing documentation (inside docs/src/main/ascoodic/elasticsearch.adoc) and maybe a small test inside the deployment module.

File: extensions/elasticsearch-rest-client/runtime/src/main/java/io/quarkus/elasticsearch/restclient/lowlevel/runtime/RestClientBuilderHelper.java
Comment (eAybars) [reply extensions/elasticsearch-rest-client/runtime/src/main/java/io/quarkus/elasticsearch/restclient/lowlevel/runtime/RestClientBuilderHelper.java:85]: Do you mean the comment line?

**Comment by yrodiere** [comment]
Nothing to add to Loïc's review; I agree it needs documentation and some tests. Looks good apart from that.

File: extensions/elasticsearch-rest-client/runtime/src/main/java/io/quarkus/elasticsearch/restclient/lowlevel/runtime/RestClientBuilderHelper.java
Comment (loicmathieu) [reply extensions/elasticsearch-rest-client/runtime/src/main/java/io/quarkus/elasticsearch/restclient/lowlevel/runtime/RestClientBuilderHelper.java:85]: No, this one:

```
HttpAsyncClientBuilder result = httpClientBuilder;
```

I don't think copying the `httpClientBuilder` inside a `result` variable didn't adds anything here, but it's nitpicking ;)

Comment (eAybars) [reply extensions/elasticsearch-rest-client/runtime/src/main/java/io/quarkus/elasticsearch/restclient/lowlevel/runtime/RestClientBuilderHelper.java:85]: Do you suggest I update the original "httpClientBuilder" variable with the value returned from the HttpClientConfigCallback implementations within the for loop or, I should discard the return values of user supplied HttpClientConfigCallback implementations and just return the httpClientBuilder?

Comment (loicmathieu) [reply extensions/elasticsearch-rest-client/runtime/src/main/java/io/quarkus/elasticsearch/restclient/lowlevel/runtime/RestClientBuilderHelper.java:85]: Sorry, I read to quickly the code and didn't notice that you need to re-assign the httpClientBuilder to itself then return it.
OK for me then

**Comment by eAybars** [comment]
Hm, that is strange. JDK11 Windows test seams to be failing on the pipeline, but the pipeline completed successfully on my fork on Github. If I understand it correctly, It seams like a timeout issue when running `io.quarkus.resteasy.reactive.server.test.simple.SimpleQuarkusRestTestCase`

File: docs/src/main/asciidoc/elasticsearch.adoc
Comment (yrodiere) [inline docs/src/main/asciidoc/elasticsearch.adoc:417]: This section really is about listing configuration properties, it's a bit odd to add text at the bottom, and people are likely to miss it.

I would move this text to the "Configuring Elasticsearch" section higher in this document?

Comment (eAybars) [reply docs/src/main/asciidoc/elasticsearch.adoc:417]: Sure thing, should I create a new section like "Programmatically Configuring Elasticsearch" right below the "Configuring Elasticsearch" section as in the [hibernate guide](https://quarkus.io/guides/hibernate-orm), or should I just continue within the Configuring Elasticsearch section? Also I was a bit unsure whether I should include some code examples or not or maybe simply include a link reference to a [here](https://www.elastic.co/guide/en/elasticsearch/client/java-rest/current/_encrypted_communication.html) for some sample usage from elasticsearch documentation. Could you please advise on that as well?

Comment (yrodiere) [reply docs/src/main/asciidoc/elasticsearch.adoc:417]: Not sure what the rules are exactly, in some cases you have to update the quickstart as well... @loicmathieu, @gsmet , any opinion?

Comment (loicmathieu) [reply docs/src/main/asciidoc/elasticsearch.adoc:417]: Quickstart is 60/40. I don't think this customization should be there.

As for adding it to the existing or creating a new one, your choice ;)

**Comment by loicmathieu** [comment]
@eAybars we had CI issue yesterday and some PRs has been cancelled, that's why you saw many failing check.

File: docs/src/main/asciidoc/elasticsearch.adoc
Comment (eAybars) [reply docs/src/main/asciidoc/elasticsearch.adoc:417]: Any suggestions on including code samples or links as I mentioned earlier? Would it be ok if I include the "Encrypted communication" example from the Elasticsearch documentation here?

Comment (loicmathieu) [reply docs/src/main/asciidoc/elasticsearch.adoc:417]: Yes, sounds good as it's the example on the Elasticsearch doc.

Comment (eAybars) [reply docs/src/main/asciidoc/elasticsearch.adoc:417]: Cheers, I will update the document as per your suggestions.

**Comment by gsmet** [comment]
@loicmathieu @yrodiere I let you both drive this one!

**Comment by yrodiere** [comment]
Looks good to me. Thanks @eAybars!

@gsmet I don't have push access to the Quarkus repository, so I'll let you or @loicmathieu merge this PR.

**Comment by eAybars** [comment]
> Looks good to me. Thanks @eAybars!

Glad that I could help, thanks.

**Review by loicmathieu — APPROVED** [review]
LGTM

File: docs/src/main/asciidoc/elasticsearch.adoc
Comment (loicmathieu) [inline docs/src/main/asciidoc/elasticsearch.adoc:264]:
```suggestion
On top of the parametric configuration, you can also programmatically apply additional configuration to the client by implementing a `RestClientBuilder.HttpClientConfigCallback` and annotating it with `ElasticsearchClientConfig`. You may provide multiple implementations and configuration provided by each implementation will be applied in a randomly ordered cascading manner.
```

File: docs/src/main/asciidoc/elasticsearch.adoc
Comment (loicmathieu) [inline docs/src/main/asciidoc/elasticsearch.adoc:266]:
```suggestion
For example, when accessing an Elasticsearch cluster that is setup for TLS on the HTTP layer, the client needs to trust the certificate that Elasticsearch is using. The following is an example of setting up the client to trust the CA that has signed the certificate that Elasticsearch is using, when that CA certificate is available in a PKCS#12 keystore.
```

**Review by loicmathieu — CHANGES_REQUESTED** [review]
I'm not a native english speaker but I think the 's' on configuration is not needed (or it should be pluralized when used later on also).

If you can just correct this I'll merge it immediatly as the CI failure is in an un-related area.

**Comment by gsmet** [comment]
This needs some adjustments (squash + a typo). I'll take care of them.

**Review by gsmet — APPROVED** [review]
OK, let's merge it and get it for CR1.

**Pull request merged**
